File: algorithms/code/leetcode/lc206_reverse_linked_list/LC206ReverseLinkedList.java
package leetcode.lc206_reverse_linked_list;

/**
 * https://leetcode.com/problems/reverse-linked-list/ #easy
 */
public final class LC206ReverseLinkedList {
    public static class ListNode {
        public int val;
        public ListNode next;

        ListNode(final int x) {
            val = x;
        }
    }

    public ListNode reverseList(final ListNode headNode) {
        ListNode head = headNode;
        if (head == null) {
            return (null);
        }
        ListNode ante = null;
        while (head != null) {
            ListNode post = head.next;
            head.next = ante;
            ante = head;
            head = post;
        }
        return ante;
    }
}
package leetcode.lc206_reverse_linked_list;

import org.junit.jupiter.api.Test;

import leetcode.lc206_reverse_linked_list.LC206ReverseLinkedList.ListNode;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

public final class LC206ReverseLinkedListTests {
    public ListNode generate(final int length) {
        ListNode list = null;
        ListNode last = null;
        for (int i = 0; i < length; i++) {
            ListNode node = new ListNode(i);
            node.next = null;
            if (list == null) {
                list = node;
            } else {
                last.next = node;
            }
            last = node;
        }
        return list;
    }

    public int[] freeze(final ListNode listNode) {
        ListNode list = listNode;
        int count = 0;
        ListNode copy = list;
        while (copy != null) {
            count++;
            copy = copy.next;
        }
        int[] frozen = new int[count];
        for (int i = 0; i < count; i++) {
            frozen[i] = list.val;
            list = list.next;
        }
        return frozen;
    }

    private void generic(final ListNode list) throws Exception {
        int[] original = freeze(list);
        int[] result = freeze(new LC206ReverseLinkedList().reverseList(list));
        assertEquals(original.length, result.length);
        for (int i = 0; i < original.length; i++) {
            assertEquals(original[original.length - 1 - i], result[i]);
        }
    }

    @Test
    public void test15() throws Exception {
        ListNode list = generate(15);
        generic(list);
    }

    @Test
    public void test1() throws Exception {
        ListNode list = generate(1);
        generic(list);
    }

    @Test
    public void testNothing() throws Exception {
        assertNull(new LC206ReverseLinkedList().reverseList(null));
    }
}

File: algorithms/code/leetcode/lc206_reverse_linked_list/lc206_reverse_linked_list.rs
// https://leetcode.com/problems/reverse-linked-list/
// #easy

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

pub struct Solution;

impl Solution {
    pub fn reverse_list(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut ante: Option<Box<ListNode>> = None;
        while let Some(mut post) = head {
            head = post.next.take();
            post.next = ante;
            ante = Some(post);
        }
        ante
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn vector_to_list(vector: Vec<i32>) -> Option<Box<ListNode>> {
        let mut list: Option<Box<ListNode>> = None;
        for &value in vector.iter().rev() {
            let node = Box::new(ListNode {
                val: value,
                next: list,
            });
            list = Some(node);
        }
        list
    }

    fn list_to_vector(mut list: Option<Box<ListNode>>) -> Vec<i32> {
        let mut vector = Vec::new();
        while let Some(node) = list {
            vector.push(node.val);
            list = node.next;
        }
        vector
    }

    #[test]
    fn test_15() {
        let vector: Vec<i32> = (0..15).collect();
        let list = vector_to_list(vector);
        let result = Solution::reverse_list(list);
        let expected: Vec<i32> = (0..15).rev().collect();
        assert_eq!(expected, list_to_vector(result));
    }

    #[test]
    fn test_1() {
        let list = vector_to_list(vec![0]);
        let result = Solution::reverse_list(list);
        assert_eq!(vec![0], list_to_vector(result));
    }

    #[test]
    fn test_nothing() {
        assert!(Solution::reverse_list(None).is_none());
    }
}
